mcstatus/__main__.py

```diff
@@ -1,75 +1,153 @@
+
+    java_res = response if isinstance(response, JavaStatusResponse) else None
+
+    if not java_res:
+        player_sample = ""
+    elif java_res.players.sample is not None:
+        player_sample = str([f"{player.name} ({player.id})" for player in java_res.players.sample])
     else:
         player_sample = "No players online"
 
-    print(f"version: v{response.version.name} (protocol {response.version.protocol})")
-    print(f'motd: "{response.motd}"')
-    print(f"players: {response.players.online}/{response.players.max} {player_sample}")
+    if player_sample:
+        player_sample = " " + player_sample
 
+    print(f"version: {_kind(server)} {response.version.name} (protocol {response.version.protocol})")
+    print(f"motd:{_motd(response.motd)}")
+    print(f"players: {response.players.online}/{response.players.max}{player_sample}")
+    print(f"ping: {response.latency:.2f} ms")
+    return 0
 
-def json(server: JavaServer) -> None:
-    data = {}
-    data["online"] = False
-    # Build data with responses and quit on exception
+
+def json(server: SupportedServers) -> int:
+    data = {"online": False, "kind": _kind(server)}
+
+    status_res = query_res = None
     try:
         status_res = server.status(tries=1)
-        data["version"] = status_res.version.name
-        data["protocol"] = status_res.version.protocol
-        data["motd"] = status_res.motd.raw
-        data["player_count"] = status_res.players.online
-        data["player_max"] = status_res.players.max
-        data["players"] = []
-        if status_res.players.sample is not None:
-            data["players"] = [{"name": player.name, "id": player.id} for player in status_res.players.sample]
-
-        data["ping"] = status_res.latency
-        data["online"] = True
-
-        query_res = server.query(tries=1)  # type: ignore[call-arg] # tries is supported with retry decorator
-        data["host_ip"] = query_res.raw["hostip"]
-        data["host_port"] = query_res.raw["hostport"]
-        data["map"] = query_res.map
-        data["plugins"] = query_res.software.plugins
-    except Exception:  # TODO: Check what this actually excepts
-        pass
-    print(json_dumps(data))
-
-
-def query(server: JavaServer) -> None:
+        if isinstance(server, JavaServer):
+            query_res = server.query(tries=1)
+    except Exception as e:
+        if status_res is None:
+            data["error"] = str(e)
+
+    # construct 'data' dict outside try/except to ensure data processing errors
+    # are noticed.
+    data["online"] = bool(status_res or query_res)
+    if status_res is not None:
+        data["status"] = dataclasses.asdict(status_res)
+
+        # XXX: hack to fixup MOTD serialisation. proper JSON serialisation for the Motd class should be implemented elsewhere.
+        assert "motd" in data["status"]
+        data["status"]["motd"] = {"raw": status_res.motd.to_minecraft()}
```

Comment:
I promise, this is the last comment I am planning to leave.

Sorry for a bit confusing implementation of our MOTD system, but before the output goes to the end user, it should be simplified:

```
# ms.minerival.org, `to_minecraft()` vs `simplify().to_minecraft()`
                §r§4§lMineRival §r§7§l| §r§fBack In Action!\n   §r§a§lSEASON 3 RELEASE DATE §r§7§l→ §r§6§lJANUARY 27th!§r
                §4§lMineRival §7| §fBack In Action!\n   §aSEASON 3 RELEASE DATE §7→ §6JANUARY 27th!                
```

This is not done automatically, because `.simplify` is too slow to run it on every request.

```suggestion
        data["status"]["motd"] = {"raw": status_res.motd.simplify().to_minecraft()}
```

(this also applies to another usage of MOTD in this file)

Reply:
I've made the change. I'm not sure what other usage you mean, since the other usages are `to_ansi` and the redundant symbols wont be visible to the user anyway.

Reply:
> the redundant symbols wont be visible to the user after their terminal displays the colours.

True

> I also think the simplifier could be improved to run in linear time, but I leave this to you.

We have tried, and [failed](https://github.com/py-mine/mcstatus/pull/335#discussion_r1051658497). If you have an idea, try to open a PR (but you must not break any tests)

> Now thinking about it more, I don't think it's appropriate to simplify the MOTD in a field called raw. One would assume it's directly from the server.

Yeah, maybe not naming this `raw` is the proper solution?

Reply:
I've changed the JSON field names. Re PR 306, there are very many comments and I don't know which parts you're trying to point out to me.

Reply:
Here is a link that works: https://github.com/py-mine/mcstatus/pull/335#discussion_r1053587201 (look at one comment before linked).

Basically, we have to reiterate over Motd a few times, to catch some things like

```py
        obj = Motd([Formatting.BOLD, "", Formatting.RESET, "", MinecraftColor.RED, ""])
>       assert obj.simplified() == Motd([])
E       AssertionError: assert Motd(parsed=[<Formatting.BOLD: 'l'>, <Formatting.RESET: 'r'>, <MinecraftColor.RED: 'c'>], bedrock=False) == Motd(parsed=[], bedrock=False)
```
(TLDR ItsDrike proposed `O(n)` solution, which didn't work because of this test)

---

Regarding "not naming this `raw` is the proper solution", I meant:

```diff
-data["status"]["motd"] = {"raw": status_res.motd.simplify().to_minecraft()}
+data["status"]["motd"] = status_res.motd.simplify().to_minecraft()
```

I don't see a reason to have both unsimplified and simplified MOTD in the response, as those are basically the same (but simplified has much less noise)

Reply:
I think that, for a server owner who just wants to see their MOTD is set correctly, it would be unexpected for a status tool to do any modification of the response. This is why I've marked it clearly.

Re the simplified code itself, I see you two have already considered it at length. I'll leave it.

Reply:
> unexpected for a status tool to do any modification of the response.

Unfortunately, we cannot guess how MOTD was written in server.properties, as it is transformed into another format on server side.

Simplification is only needed because of specifics of our parser; it creates a bunch of noise, and we cannot remove it without looking at the complete elements list (aka `Motd.parsed`). Thus, we cannot put simplifying process into parsing process, and unable to do simplifying fast enough to run on every server request.

Here is a good example how unsimplified and simplified MOTDs are different: https://github.com/py-mine/mcstatus/compare/a463a49c94c0b2bb9e8cb47e036db772f3c8c477..d19f58699f37a40cb45dc4c4aa435ce05de8c2a6